### PR TITLE
[PR2] Bin-pack delete files flag, MoR feature validation (only spark3.5), and observability

### PR DESCRIPTION
**Pull request by cbb330**
## Summary
To support row level updates prevalent in CDC use cases, we will enable MoR in OpenHouse. Enabling MoR has a risk of delete file accumulation which affects storage quota utilization and compute performance. Therefore we must mitigate this risk with efficient compaction of delete files.

Here we are implementing:
- assertions that delete file compaction is enabled in the OpenHouse stack, indirectly with spark3.5 due to spark3.1 inability to create delete files
- asserting feature gaps which we have researched about delete file compaction
- key observability metrics for this delete file compaction
- exposing the flag to trigger this compaction

minimal tests to be added:
```java
testDeleteFilesCanBeCreated();
testCompactionCanRemovePositionDeleteFiles();
testCompactionCanRemoveEqualityDeleteFiles();
```

tests which I will add in a PR3. these aim to reinforce our foundational knowledge or cover known edge cases:
```java
// this validates our understanding of datafile sequence number being the primary mechanism to remove delete files as is documented here: https://docs.google.com/document/d/11d-cIUR_89kRsMmWnEoxXGZCvp7L4TUmPJqUC60zB5M/edit?tab=t.0#bookmark=id.a485992lwbxz 
testSnapshotExpirationCanRemovePositionDeleteFiles(); 
testSnapshotExpirationCanRemoveEqualityDeleteFiles(); 

// these are documented issues where delete files may accumulate, needs understanding of data sequence number
testDeleteFilesNotExpirableBecauseOldDatafile();
testDeleteFilesNotExpirableBecauseDangling();

// this is our recovery process to implement in the future  via spark application in case delete files do accumulate from above
testProcessToCompactNonExpirableDeleteFiles();
```

[Issue](https://github.com/linkedin/openhouse/issues/#nnn)] Briefly discuss the summary of the changes made in this 
pull request in 2-3 lines.

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [x] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [ ] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/MinimalSparkMoRTest.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        .createDataset(records, Encoders.bean(SimpleRecord.class))
+        .coalesce(1)
+        .writeTo(tableName)
+        .append();
+
+    sql("DELETE FROM %s WHERE id = 1 and data='a'", tableName);
+    sql("DELETE FROM %s WHERE id = 2 and data='d'", tableName);
+    sql("DELETE FROM %s WHERE id = 1 and data='c'", tableName);
+
+    // TODO: put this in the tablestats API
```
Comment (autumnust) [inline apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/MinimalSparkMoRTest.java:124]: what does this todo mean?

File: tables-test-fixtures/tables-test-fixtures-iceberg-1.2/src/main/java/com/linkedin/openhouse/tablestest/OpenHouseSparkITest.java [hunk 2]
```diff
@@ -23,7 +23,7 @@ public class OpenHouseSparkITest {
    *
    * <p>Do not {@link SparkSession#close()} a created SparkSession.
    */
-  protected synchronized SparkSession getSparkSession() throws Exception {
+  protected static synchronized SparkSession getSparkSession() throws Exception {
```
Comment (autumnust) [inline tables-test-fixtures/tables-test-fixtures-iceberg-1.2/src/main/java/com/linkedin/openhouse/tablestest/OpenHouseSparkITest.java:26]: why and what's the implication of this change?

**Review by autumnust — CHANGES_REQUESTED** [review]

File: apps/spark-3.5/build.gradle [hunk 3]
```diff
@@ -16,6 +16,9 @@ ext {
     tablesTestFixturesModule = ":tables-test-fixtures:tables-test-fixtures-iceberg-1.5_2.12"
 }
 
+configurations.all {
+    exclude group: 'org.apache.iceberg', module: 'iceberg-data'
```
[the remark this answers is not in the document]
Comment (cbb330) [reply apps/spark-3.5/build.gradle:20]: i tried this and 2 or 3 submodules e.g. libs:datalayout, services:common, we're also bringing in iceberg-data

this is a quick workaround. later i will try to not use this block, but it may still become more readable/understandable to have this block vs excluding iceberg-data from several sources

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/MinimalSparkMoRTest.java (hunk 1 above)
Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/MinimalSparkMoRTest.java:124]: elsewhere in compaction test code is assertions based on the tableStats utility method.

since we plan to use tableStats as the monitoring method for delete file accumulation, i will do the same thing in these unittests. knocking two birds with one stone (validating delete file feature AND implementing the monitoring for it)

**Comment by autumnust** [comment]
@cbb330  is it ready for review or still wip?

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+    // this asserts the dangling delete problem, since compaction did NOT remove the final delete
+    IcebergTableStats stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumPositionDeleteFiles()).isEqualTo(3L);
+    assertThat(stats.getNumCurrentSnapshotPositionDeleteFiles()).isEqualTo(1L);
+
+    // notice that running compaction again with no new commits cleans up the dangling delete
+    rewriteFunc.apply(ops, table);
+    stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumPositionDeleteFiles()).isEqualTo(3L);
```
Comment (autumnust) [inline apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:143]: this represents the number of pos-deletes across the whole tables' history?

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    assertThat(stats.getNumCurrentSnapshotPositionDeleteFiles()).isEqualTo(0L);
```
Comment (autumnust) [inline apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:144]: I thought `rewrite_position_delete_files` running after `rewrite_data_files` is required to have this. Did we discover something different after we discussed this initially?

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+    try (EqualityDeleteWriter<Record> clsEqDeleteWriter = eqDeleteWriter) {
+      clsEqDeleteWriter.write(deleteRecord);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    table.newRowDelta().addDeletes(eqDeleteWriter.toDeleteFile()).commit();
+  }
+
+  private EncryptedOutputFile createEncryptedOutputFile(
```
Comment (autumnust) [inline apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:220]: For my understanding: is encryption a related context here or this is mainly from Iceberg's testing code ?

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/spark/DataCompactionSparkApp.java [hunk 7]
```diff
@@ -20,7 +20,7 @@
  * <p>Example of invocation: com.linkedin.openhouse.jobs.spark.DataCompactionSparkApp --tableName
  * db.testTable --targetByteSize 1048576 --maxByteSizeRatio 0.75 --maxByteSizeRatio 1.8
  * --minInputFiles 5 --maxConcurrentFileGroupRewrites 2 --partialProgressEnabled
- * --partialProgressMaxCommits 10
+ * --partialProgressMaxCommits 10 --deleteFileThreshold 0
```
Comment (autumnust) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/spark/DataCompactionSparkApp.java:23]: hopefully 0 as default is safe?

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/DataFilesSummary.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  private Long sumOfFileSizeBytes;
+  private Long totalFileCount;
+
+  /**
+   * Returns the Spark Encoder for this class.
+   *
+   * @return Encoder for ExampleData
+   */
+  public static Encoder<DataFilesSummary> getEncoder() {
+    return Encoders.bean(DataFilesSummary.class);
```
Comment (autumnust) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/DataFilesSummary.java:24]: how costly is this method -- shall we make Encoder object static and lazily init it only once ?

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java [hunk 9]
```diff
@@ -79,11 +115,38 @@ public static IcebergTableStats populateStatsOfAllReferencedFiles(
   public static IcebergTableStats populateStatsForSnapshots(
       String fqtn, Table table, SparkSession spark, IcebergTableStats stats) {
 
-    Dataset<Row> dataFiles = getAllDataFilesCount(table, spark, MetadataTableType.FILES);
-    long countOfDataFiles = dataFiles.count();
-
-    // Calculate sum of file sizes in bytes in above allDataFiles
-    long sumOfFileSizeBytes = getSumOfFileSizeBytes(dataFiles);
+    // TODO: turn these longs into attributes of a class and use the class as a schema which spark
```
Comment (autumnust) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:118]: is this stale since you already have `DataFilesSummary` ?

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java [hunk 10]
```diff
@@ -190,22 +264,23 @@ private static long getManifestFilesCount(
   }
 
   /** Get all data files count depending on metadata type to query. */
-  private static Dataset<Row> getAllDataFilesCount(
+  private static Map<Integer, DataFilesSummary> getFileMetadataTable(
       Table table, SparkSession spark, MetadataTableType metadataTableType) {
-    return SparkTableUtil.loadMetadataTable(spark, table, metadataTableType)
-        .selectExpr(new String[] {"file_path", "file_size_in_bytes"})
-        .dropDuplicates("file_path", "file_size_in_bytes");
```
Comment (autumnust) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:197]: is duplication of `file_path` not necessary? Seems this is no longer there in the new implementation

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java [hunk 11]
```diff
@@ -190,22 +264,23 @@ private static long getManifestFilesCount(
   }
 
   /** Get all data files count depending on metadata type to query. */
```
Comment (autumnust) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:266]: this javadoc requires fix

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+    // first run of compaction
+    RewriteDataFiles.Result result = rewriteFunc.apply(ops, table);
+    Assertions.assertEquals(1, result.addedDataFilesCount());
+    Assertions.assertEquals(1, result.rewrittenDataFilesCount());
+
+    // this asserts the dangling delete problem, since compaction did NOT remove the final delete
+    IcebergTableStats stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumPositionDeleteFiles()).isEqualTo(3L);
+    assertThat(stats.getNumCurrentSnapshotPositionDeleteFiles()).isEqualTo(1L);
```
Comment (autumnust) [inline apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:138]: for my understanding: why is there still a pos-delete file here as part of the current snapshot; also why is this pos-delete, which process generates this and how does that process determine between pos vs. eq?

**Review by autumnust — CHANGES_REQUESTED** [review]

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java (hunk 4 above)
Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:143]: yes. as opposed to the current snapshot metrics will only have latest table history

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/spark/DataCompactionSparkApp.java (hunk 7 above)
Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/spark/DataCompactionSparkApp.java:23]: default is MAX_INT,

we will pass 0 in as an override in short term, via kube cli parameters. it is safe for 12 hour cadence and for minimal tables.

longer term will require a more pragmatic number based on usage. or else, a number derived from DLO strategy

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java (hunk 12 above)
Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:138]: > why is there still a pos-delete file here as part of the current snapshot

delete here is dangling due to compaction generating a new datafile with the same data sequence number as the delete file. because the resulting datafile and delete file have same sequence number, the delete file isn't chosen to be dropped.

https://github.com/apache/iceberg/issues/4127


> why is this pos-delete

engine specific decision to create position delete. as far as i know, spark does not actually create equality deletes at all (only flink, and kafka connect)

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java (hunk 6 above)
Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:220]: from iceberg's testing code,

i. tried to find a similar method without encryption since we don't need it and it is a bit of a distraction, but there doesn't seem to be one.

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java (hunk 10 above)
Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:197]: i don't see that it's necessary, not seeing online that this table can contain dupes. i also don't see context on why it was added in the original blame.

@maluchari maybe you have more context?

i will add it back though to be sure

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java (hunk 11 above)
Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:266]: fixed

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/DataFilesSummary.java (hunk 8 above)
Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/DataFilesSummary.java:24]: not sure but good call. will share the object to be safe

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java (hunk 9 above)
Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:118]: yep

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java (hunk 5 above)
Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:144]: yes, i thought the same. I assumed the dangling delete logic would still apply: compaction installs a new snapshot with the same data sequence number as the delete and thus the delete cannot be pruned.

 I had previously thought that a new data sequence number (a new snapshot with data) would need to be attached to the table and then a full table rewrite. 


what is happening in this test appears to conflict with the developer notes here:
https://docs.google.com/document/d/11d-cIUR_89kRsMmWnEoxXGZCvp7L4TUmPJqUC60zB5M/edit?tab=t.0#bookmark=id.cyzox4b19g6s


Perhaps we may be doing something wrong if we don't fully understand the results, in this case I plan to follow up with a more detailed validatino of the data sequence number and snapshot validation -- but due to time will focus on providing the feature as-is first.

File: tables-test-fixtures/tables-test-fixtures-iceberg-1.2/src/main/java/com/linkedin/openhouse/tablestest/OpenHouseSparkITest.java (hunk 2 above)
Comment (cbb330) [reply tables-test-fixtures/tables-test-fixtures-iceberg-1.2/src/main/java/com/linkedin/openhouse/tablestest/OpenHouseSparkITest.java]: removed this code

File: apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java (hunk 12 above)
Comment (autumnust) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:138]: for the first question: but why does delete files become one although metadata-compaction is not called?

Comment (cbb330) [reply apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java:143]: compaction is called on line 131 above.

1. insert 3 deletes (shown in test)
2. table has 3 deletes associated with the current snapshot (this is not asserted, i will add it though for clarity)
3. compaction first time
4. only 1 delete associated with current snpashot: first 2 delete files were dropped due to being older than the newest datafile created by compaction in step 3, last 1 delete file is dangling

File: services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java [hunk 13]
```diff
@@ -36,5 +36,21 @@ public class IcebergTableStats extends BaseTableMetadata {
...
+  private Long totalEqualityDeleteFileSizeInBytes;
+
+  private Long numCurrentSnapshotPositionDeleteFiles;
+
+  private Long totalCurrentSnapshotPositionDeleteFileSizeInBytes;
+
+  private Long numCurrentSnapshotEqualityDeleteFiles;
+
+  private Long totalCurrentSnapshotEqualityDeleteFileSizeInBytes;
+
```
Comment (autumnust) [inline services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java:54]: Let's make sure these changes are safe. cc @maluchari

**Review by autumnust — CHANGES_REQUESTED** [review]
last round after discussion in person: 
- let's make sure the stats changes are safe in the internal wrapper layer. 
- meanwhile, we should get more understanding on the delete-file compaction within Spark's implementation. Will do that separately. This is not blocking though given the changes are not going to impact the existing production compaction jobs.

File: services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java (hunk 13 above)
Comment (cbb330) [reply services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java:54]: @maluchari as I read in the internal wrapper,

there is a coded set of assignments from IcebergTableStats 's  attributes -> the internal event schema.

so without changing anything in the internal wrapper, the fields in IcebergTableStats go unused. 

thus it is safe to merge this PR, and care must be taken in the internal wrapper when we update the set of assignemnts and change the event schema.

**Review by autumnust — APPROVED** [review]

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java (hunk 10 above)
Comment (maluchari) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:197]: duplicate file paths is required since same file can be referenced multiple types across snapshots.

File: services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java (hunk 13 above)
Comment (maluchari) [reply services/common/src/main/java/com/linkedin/openhouse/common/stats/model/IcebergTableStats.java:54]: This specific block of lines LGTM

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java [hunk 14]
```diff
@@ -47,28 +50,61 @@ public static IcebergTableStats populateStatsOfAllReferencedFiles(
     long totalMetadataFilesCount =
         referencedManifestFilesCount + referencedManifestListFilesCount + metadataFilesCount;
 
-    Dataset<Row> allDataFiles =
-        getAllDataFilesCount(table, spark, MetadataTableType.ALL_DATA_FILES);
-    long countOfDataFiles = allDataFiles.count();
-
-    // Calculate sum of file sizes in bytes in above allDataFiles
-    long sumOfFileSizeBytes = getSumOfFileSizeBytes(allDataFiles);
+    Map<Integer, DataFilesSummary> allDataFilesSummary =
```
Comment (maluchari) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:53]: allFilesSummary since this includes data files + delete files

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java [hunk 15]
```diff
@@ -79,11 +115,35 @@ public static IcebergTableStats populateStatsOfAllReferencedFiles(
   public static IcebergTableStats populateStatsForSnapshots(
       String fqtn, Table table, SparkSession spark, IcebergTableStats stats) {
 
-    Dataset<Row> dataFiles = getAllDataFilesCount(table, spark, MetadataTableType.FILES);
-    long countOfDataFiles = dataFiles.count();
-
-    // Calculate sum of file sizes in bytes in above allDataFiles
-    long sumOfFileSizeBytes = getSumOfFileSizeBytes(dataFiles);
+    Map<Integer, DataFilesSummary> currentSnapshotDataFilesSummary =
```
Comment (maluchari) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:118]: same here

Comment (cbb330) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/TableStatsCollectorUtil.java:118]: thanks @maluchari 

I've pushed a second PR to address this: https://github.com/linkedin/openhouse/pull/270
